### PR TITLE
Fixlenspolarizationrotation

### PR DESCRIPTION
**Pull request by briandleahy**
Fixes 371.

This fixes two bugs with the implementation of polarization rotation in the `Lens` theory. These issues were only present when the polarization was not along the x (or [1, 0]) direction. With this PR, the agreement between MieLens and Lens(Mie) does not change with polarization, and both theories produce rotationally-invariant results for rotationally-invariant objects (i.e. the parallel component of the field scattered from a sphere, evaluated at a position that is along the polarizazion vector, does not change as the polarization changes; this was always true for MieLens but was not true for Lens(Mie)).

Note that this does not change the calculations at all when the polarization is along the x-direction, so #350 is still open.

### CODE REVIEW

**Comment by briandleahy** [comment]
@ralex0 -- I made some slight edtis to the `numexpr` usage, but I don't have `numexpr` working on my machine. Can you check that the `numexpr` and no-numexpr pathwasy produce the same result?
@barkls -- as we discussed on Thurs, I also threw in 2 minor fixes with this PR (faster test_interface and making the DDA tests pass). However I'm not sure if the DDA tolerances should be changed. If you want, we can leave the DDA tests as failing for this PR. Let me know.

**Review by barkls — APPROVED** [review]

**Comment by barkls** [comment]
Looks good, I just have a few additional comments:

1) It would be nice to have a short summary in this thread of what the bugs were for future reference, especially since the minor rewrite of the Lens class makes it hard to pinpoint from the code.

2) Please update the changelog.

File: holopy/scattering/tests/test_lens.py
Comment (briandleahy) [reply holopy/scattering/tests/test_lens.py:70]: The old quadrature was optimized for arbitrary functions on an interval, using Gauss-Legendre points and weights. However, these are not optimal for periodic functions, and the actual integrals we are evaluating are periodic with period 2 pi. So I optimized the phi quadrature for a function with period 2 pi, which gives a 3x improvement in accuracy with the same computational cost.

However, the old function in the test, cos (pi * x), is not 2 pi - periodic, and therefore the quadrature does not perform as well on this function as it did before, and the test did not pass. This is a problem with the test and not the code, so I changed the test. I could have changed it to just cos(x), but  that is a more well-behaved function. Since I needed to change the function, I figured that I might as well make the test more challenging as well.

File: holopy/scattering/theory/dda.py
Comment (briandleahy) [reply holopy/scattering/theory/dda.py:85]: For some reason the `SuppressOutput`still causes crashes in ipython terminals on my machine, and this was stopping me from running DDA in an ipython shell.  At some point I should come back to the `SuppressOutput`ipython crash.

**Comment by briandleahy** [comment]
> 1. It would be nice to have a short summary in this thread of what the bugs were for future reference, especially since the minor rewrite of the Lens class makes it hard to pinpoint from the code.

There were two separate problems, both just involving polarization rotations. 

The first problem was that the code rotated the object (by rotating the calculation of the scattering matrices) when the polarization was rotated. This was a simple bugfix.

The second problem was a more subtle error, corresponding to a math error in rotating the polarization (so apologies for having to read it in markdown). Take a look at Eq. 1 from the perspective paper below, which is correct for x-polarized light:
Screenshot from 2020-10-26 11-38-36
Previously, the `Lens` theory was calculating the scattered fields by assuming that the component of the scattered field parallel to the polarization was the same, independent of whether the light was x-polarized or polarized in an arbitrary direction. This is equivalent to changing every "x-hat" in Eq. (1) into a "p-hat", where "p-hat" is the polarization direction (and "y-hat" with a vector perpendicular to the polarization). This is also incorrect, because the "x-hat"s appear for two separate reasons. The terms **theta** * **S** * **x** and **phi** * **S** * **x** capture the effects of the polarization of the incoming beam, and need to rotate when the polarization changes. But the terms (**x** cos(phi) + **y** sin(phi)) etc describe theta and phi unit vectors on the lens pupil, and do not rotate when the polarization changes. This is a conceptual difference, but it ends up being a small code change at the end -- the integrands just get a cos(phi - pol_angle) and sin(phi - pol_angle) instead of cos(phi) and sin(phi), lines 141-142 and 151-152. 

If you want I can quickly TeX up the derivation and place it somewhere, perhaps here or on the group wiki.

File: holopy/scattering/tests/test_lens.py
Comment (briandleahy) [reply holopy/scattering/tests/test_lens.py]: Fixed by 4c63307b

File: holopy/scattering/tests/test_lens.py
Comment (briandleahy) [reply holopy/scattering/tests/test_lens.py]: Fixed by 4c63307b

**Comment by briandleahy** [comment]
> 2. Please update the changelog.

I've also added a line to the changelog, in 4c63307b

File: holopy/scattering/tests/test_lens.py
Comment (barkls) [reply holopy/scattering/tests/test_lens.py:70]: thanks for clarifying!

**Comment by briandleahy** [comment]
@barkls -- let me know if I missed anything or if this is ready to merge.

**Comment by barkls** [comment]
I think we're just waiting on a test of numexpr from @ralex0

**Comment by briandleahy** [comment]
@barkls -- I installed `numexpr` and checked this on my local machine. There were some issues, which I fixed. Those changes are now pushed (primarily in 7e63fc92 ). Please take a look and let me know if something seems off!

I did check that changing the `numexpr` expressions causes the tests to fail. I also adjusted the tests to compare the numexpr pathway against the numpy pathway, which in my opinion is what matters anyways.

File: holopy/scattering/tests/test_lens.py
Comment (barkls) [inline holopy/scattering/tests/test_lens.py:160]: I think this is a sign that some functionality is missing - e.g. calling Lens with a `computation_method='auto'` that can be overridden to specify numpy or numexpr. 

However, we also want to avoid complexifying [not a word] the interface, and there's no real reason you'd want to enforce numpy outside of testing so this seems fine.

Comment (barkls) [reply holopy/scattering/tests/test_lens.py:160]: I notice you do this a few times through this test file. You could consider writing a context manager for the monkey patch. That would also make it less likely someone breaks things later since the numexpr adjustments would be removed from the tests themselves.

**Review by barkls — APPROVED** [review]

**Comment by barkls** [comment]
Thanks for following up on this @briandleahy and patching things to work with numexpr. I suggested a context manager to handle the numexpr monkey-patches, but we can merge in as is if you don't feel like doing that.

At some point it would be nice to get things like DDA and numexpr tested on Travis, but that's definitely beyond the scope of this PR.

File: holopy/scattering/tests/test_lens.py
Comment (briandleahy) [reply holopy/scattering/tests/test_lens.py]: Should be fixed by babd9bcd. Let me re-run the tests quickly on my computer with numexpr working before merging this in though.

**Comment by briandleahy** [comment]
Tests work with new commit. I'm ready to merge this in when you are.

**Comment by barkls** [comment]
good fix!

**Pull request merged**
